src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -196,69 +189,94 @@ private async ValueTask ProcessEventChangeAsync(ImmutableSegmentedList<bool> cha
             /// </param>
             private async Task RecomputeTagsAsync(bool highPriority, CancellationToken cancellationToken)
             {
+                var isVisible = true;
+                var spansToTag = ImmutableArray<DocumentSnapshotSpan>.Empty;
+                SnapshotPoint? caretPosition = null;
+                ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> oldTagTrees = null!;
+                object? oldState = null;
+                TextChangeRange? textChangeRange = null;
+
+                // Enqueue work to a queue that will all tagger main thread work together in the near future. This let's
+                // us avoid hammering the dispatcher queue with lots of work that causes contention.  Additionally, use
+                // a no-throw awaitable so that in the common case where we cancel before, we don't throw an exception
+                // that can exacerbate cross process debugging scenarios.
+                await _dataSource._mainThreadManager.PerformWorkOnMainThreadAsync(() =>
```

Comment:
this is the only method that uses this new 'ui thread manager'.  At hte start, it jumps to the UI thread (along with all other taggers in the same batch) and grabs the data it needs.  then it goes to the BG to do all the work.  Finally, it goes back to the UI thread to update UI state (Again, being batched with any other taggers that are also finished.